zebra/zebra_srv6.c

```diff
@@ -0,0 +1,331 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <zebra.h>
+
+#include "network.h"
+#include "prefix.h"
+#include "stream.h"
+#include "srv6.h"
+#include "zebra/zapi_msg.h"
+#include "zebra/zserv.h"
+#include "zebra/zebra_router.h"
+#include "zebra/zebra_srv6.h"
+#include "zebra/zebra_memory.h"
+#include "zebra/zebra_errors.h"
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <arpa/inet.h>
+#include <netinet/in.h>
+
+
+static void set_seg6local_end(struct prefix_ipv6 *prefix)
+{
+	struct route_entry *re;
+	struct nexthop_group *ng;
+	struct vrf *vrf;
+	struct interface *ifp;
+	struct nexthop *nexthop;
+	int ret;
+	ifindex_t ifindex;
+
+	re = XCALLOC(MTYPE_RE, sizeof(struct route_entry));
+	re->type = ZEBRA_ROUTE_STATIC;
+	re->instance = 0;
+	re->uptime = monotime(NULL);
+	re->vrf_id = VRF_DEFAULT;
+	re->table = RT_TABLE_MAIN;
+	SET_FLAG(re->flags, ZEBRA_FLAG_SEG6LOCAL_ROUTE);
+
+	ifindex = 1;
+	vrf = vrf_lookup_by_id(VRF_DEFAULT);
+	FOR_ALL_INTERFACES (vrf, ifp) {
+		if (ifp->ifindex == 1)
+			continue;
+		if (!if_is_operative(ifp))
+			continue;
+		ifindex = ifp->ifindex;
+		break;
+	}
+
+	if (ifindex == 1)
+		zlog_err("%s: no available interface on default-vrf",
+			 __func__);
+
+	ng = nexthop_group_new();
+	nexthop = nexthop_from_ifindex(ifindex, VRF_DEFAULT);
+	if (!nexthop) {
+		flog_warn(
+			EC_ZEBRA_NEXTHOP_CREATION_FAILED,
+			"%s: Nexthops Specified: but we failed to properly create one",
+			__PRETTY_FUNCTION__);
+		nexthop_group_delete(&ng);
+		XFREE(MTYPE_RE, re);
+		return;
+	}
+
+	nexthop_add_seg6local(nexthop, ZEBRA_SEG6_LOCAL_ACTION_END, NULL);
+	nexthop_group_add_sorted(ng, nexthop);
+
+	ret = rib_add_multipath(AFI_IP6, SAFI_UNICAST,
+				(struct prefix *)prefix, NULL, re, ng);
+	if (ret < 0)
+		zlog_err("%s: can't add RE entry to rib", __func__);
+}
+
+static void locator_init(const struct srv6_locator *loc)
+{
+	size_t idx_hi, idx_lo;
+	uint16_t value = 0x0001;
+	struct prefix_ipv6 p;
+	struct srv6_function *function;
+
+	idx_hi = loc->prefix.prefixlen / 8;
+	idx_lo = idx_hi + 1;
+
+	p = loc->prefix;
+	p.prefix.s6_addr[idx_hi] = (0xff00 & value) >> 8;
+	p.prefix.s6_addr[idx_lo] = (0x00ff & value);
+	p.prefixlen = loc->prefix.prefixlen + loc->function_bits_length;
+	set_seg6local_end(&p);
+
+	function = srv6_function_alloc(&p);
+	listnode_add(loc->functions, function);
+	strlcpy(function->locator_name, loc->name,
+		sizeof(function->locator_name));
+}
+
+static void get_function_addr(struct in6_addr *locator,
```

Comment:
1. this function name is way too vague
2. this function needs a doc comment because it is not at all obvious what it does

Reply:
Sorry, this line was deleted when re-write.